graphs/mis/sync_metivier_et_al/sync_metivier_et_al.go

```diff
@@ -0,0 +1,349 @@
+		Ins:        make(map[int]bool),
+		OutPhase:   0,
+		Outs:       make(map[int]bool),
+		Responsive: true,
+	}
+}
+
+func (s *state) sendBits(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Responsive {
+			bit := s.getNeighborBit(neighbor)
+			sendMessage(node, i, &bit)
+		}
+	}
+}
+
+func (s *state) receiveBits(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if !neighbor.Responsive {
+			continue
+		}
+		received := *receiveMessage(node, i) // in this stage, bits are always sent, so it is safe to dereference
+		bit := s.getNeighborBit(neighbor)
+		if bit != received { // symmetry is broken
+			neighbor.Wins[neighbor.WinPhase] = !bit // we know whether we won this phase against this neighbor
+			neighbor.WinPhase++                     // we can start the next phase of symmetry breaking
+			neighbor.BitIndex = 0
+		} else { // still symmetrical
+			neighbor.BitIndex++ // we will have to send another bit
+		}
+	}
+}
+
+func (s *state) sendIns(node lib.Node) {
+	in := s.canEnterMIS()
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Responsive {
+			sendMessage(node, i, in)
+		}
+	}
+	if in != nil {
+		s.InsSent = true
+		s.Chosen = *in
+	}
+}
+
+func (s *state) receiveIns(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if !neighbor.Responsive {
+			continue
+		}
+		in := receiveMessage(node, i)
+		if in != nil {
+			neighbor.Ins[neighbor.InPhase] = *in
+			neighbor.InPhase++
+		}
+	}
+}
+
+func (s *state) sendOuts(node lib.Node) {
+	out := s.shouldDisconnect()
+	for i, neighbor := range s.Neighbors {
+		if neighbor.Responsive {
+			sendMessage(node, i, out)
+		}
+	}
+	if out != nil {
+		s.OutsSent = true
+		s.Done = *out
+	}
+}
+
+func (s *state) receiveOuts(node lib.Node) {
+	for i, neighbor := range s.Neighbors {
+		if !neighbor.Responsive {
+			continue
+		}
+		out := receiveMessage(node, i)
+		if out != nil {
+			neighbor.Responsive = !*out
+			neighbor.Outs[neighbor.OutPhase] = *out
+			neighbor.OutPhase++
+		}
+	}
+}
+
+func (s *state) getNeighborBit(neighbor *neighbor) bool {
+	return s.getBit(neighbor.WinPhase, neighbor.BitIndex)
+}
+
+func (s *state) getBit(phase int, index int) bool {
+	bits := s.Bits[phase]
+	if index > len(bits) {
+		panic("Earlier bits have not been drawn yet")
+	}
+	if index == len(bits) {
+		rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+		bits = append(bits, rng.Intn(2) == 1)
+	}
+	s.Bits[phase] = bits
+	return bits[index]
+}
+
+func (s *state) canEnterMIS() *bool {
+	if s.InsSent {
+		return nil
+	}
+	all := true
+	enter := true
+	for _, neighbor := range s.Neighbors {
+		win, ok := neighbor.Wins[s.Phase]
+		if !ok {
+			all = false
+		} else if !win {
+			enter = false
+		}
+	}
+	if !all {
+		return nil
+	}
+	return &enter
+}
+
+func (s *state) shouldDisconnect() *bool {
+	if s.OutsSent {
+		return nil
+	}
+	all := true
+	disconnect := false
+	for _, neighbor := range s.Neighbors {
+		in, ok := neighbor.Ins[s.Phase]
+		if !ok {
+			all = false
```

Comment:
Again, consider `return nil` and removing all

Reply:
Done